### PR TITLE
[plugin-accessibility] Add integration with Axe Core

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        <script src="../../webjars/bootstrap/5.3.0-alpha3/js/bootstrap.min.js"></script>
+        <script src="../../webjars/jquery/3.6.4/jquery.min.js"></script>
+        <link
+            rel="stylesheet"
+            href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/styles/stackoverflow-light.min.css"
+        />
+        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/highlight.min.js"></script>
+        <link
+            rel="icon"
+            href="https://www.deque.com/wp-content/uploads/2018/03/cropped-DQ_SecondaryLogo_HeroBlue_RGB-1-32x32.png"
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl:115]: why is it needed?

File: vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        </style>
+        <link rel="stylesheet" href="../../styles.css"/>
+        <link rel="stylesheet" href="../../webjars/vividus/buttons.css"/>
+        <script src="../../webjars/bootstrap/5.3.0-alpha3/js/bootstrap.min.js"></script>
+        <script src="../../webjars/jquery/3.6.4/jquery.min.js"></script>
+        <link
+            rel="stylesheet"
+            href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/styles/stackoverflow-light.min.css"
+        />
+        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/highlight.min.js"></script>
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl:112]: we shouldn't use external resources in report

File: vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/code.ftl [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            checkOptions.getRowsAs(AccessibilityCheckOptions.class)
+                        .forEach(options -> {
+                            List<AccessibilityViolation> violations = accessibilityTestEngine
+                                    .execute(accessibilityEngine, options, AccessibilityViolation.class);
+                            ViolationLevel level = options.getLevel();
+                            publishAttachment(currentUrl, violations);
+                            softAssert.assertThat(String.format(MESSAGE, level, currentUrl),
+                                violations.stream().filter(v -> v.getTypeCode() <= level.getCode()).count(), equalTo(0L));
+                        });
+        }
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/code.ftl:26]: ?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-accessibility.adoc [hunk 4]
```diff
@@ -22,13 +20,66 @@ _Deprecated syntax (will be removed in VIVIDUS 0.7.0)_:
...
+----
+accessibility.engine=AXE_CORE
+----
+
+Accessibility test options:
+
+* `$options` - The accessibility validation options. Where:
+** [subs=+quotes]`*standard*` - The tag of accessibility standard to check a page against, the list of available tags can be found https://www.deque.com/axe/core-documentation/api-documentation/#axe-core-tags[here].
+** `elementsToCheck` - The comma-separated xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to verify.
+** `elementsToIgnore` - The comma-separated xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to ignore.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-accessibility.adoc:42]:
```suggestion
** `elementsToCheck` - The comma-separated list of xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to verify.
** `elementsToIgnore` - The comma-separated list of xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to ignore.
```

File: docs/modules/plugins/pages/plugin-accessibility.adoc [hunk 5]
```diff
@@ -22,13 +20,66 @@ _Deprecated syntax (will be removed in VIVIDUS 0.7.0)_:
...
+** `violationsToIgnore` - The comma-separated list of https://dequeuniversity.com/rules/axe/{axe-version}/[violations] to ignore.
+** `violationsToCheck` - The comma-separated list of https://dequeuniversity.com/rules/axe/{axe-version}/[violations] to check.
+
+.Check page against WCAG 2.2 Level AA
+[source,gherkin]
+----
+Given I am on page with URL `https://vividus-test-site.onrender.com/`
+Then I test accessibility:
+|standard|
+|wcag22aa|
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-accessibility.adoc:52]:
```suggestion
|WCAG22AA|
```

File: vividus-allure-adaptor/src/main/resources/allure-customization/webjars/js-beautify/1.13.4/styles/stackoverflow-light.min.css [hunk 6]
```diff
@@ -0,0 +1,13 @@
+pre code.hljs{display:block;overflow-x:auto;padding:1em}code.hljs{padding:3px 5px}/*!
+  Theme: StackOverflow Light
+  Description: Light theme as used on stackoverflow.com
```
Comment (valfirst) [inline vividus-allure-adaptor/src/main/resources/allure-customization/webjars/js-beautify/1.13.4/styles/stackoverflow-light.min.css:3]: why is it needed?

File: docs/modules/plugins/pages/plugin-accessibility.adoc [hunk 7]
```diff
@@ -22,13 +20,66 @@ _Deprecated syntax (will be removed in VIVIDUS 0.7.0)_:
...
+
 * `$options` - The accessibility validation options. Where:
 ** [subs=+quotes]`*standard*` -  The accessibility stanadard to verify against. One of `WCAG2A`, `WCAG2AA`, `WCAG2AAA`, `Section 508`
 ** [subs=+quotes]`*level*` - The violation level used to check. One of:
 *** `ERROR` - only errors will be validated,
 *** `WARNING` - error and warning will be validated,
 *** `NOTICE` - all the levels will be valiadted.
-** `locators` - The comma-separated xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to verify
+** `elementsToCheck` - The comma-separated xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to verify
 ** `elementsToIgnore` - The comma-separated xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to ignore
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-accessibility.adoc:83]:
```suggestion
** `elementsToCheck` - The comma-separated list of xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to verify
** `elementsToIgnore` - The comma-separated list of xref:vividus:plugins:plugin-web-app.adoc#_locator[locators] of elements to ignore
```

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 8]
```diff
@@ -83,21 +90,54 @@ public AccessibilitySteps(AccessibilityTestEngine accessibilityTestEngine,
...
+
+                attachmentPublisher.publishAttachment(
+                    "/org/vividus/accessibility/axe-accessibility-report.ftl",
+                    Map.of("entries", report, "url", currentUrl, "run", options.getRunOnly()),
+                    "Accessibility report for page: " + currentUrl
+                );
+
+                long failures = report.stream().filter(e -> ResultType.FAILED == e.getType())
+                        .map(AxeReportEntry::getResults).map(List::size).findFirst().orElse(0);
+                softAssert.assertThat("Accessibility check for page: " + currentUrl, failures, equalTo(0L));
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:124]: the assertion message is not clear enough

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 9]
```diff
@@ -83,21 +90,54 @@ public AccessibilitySteps(AccessibilityTestEngine accessibilityTestEngine,
...
+                publishAttachment(currentUrl, violations);
+                softAssert.assertThat(String.format(MESSAGE, level, currentUrl),
+                    violations.stream().filter(v -> v.getTypeCode() <= level.getCode()).count(), equalTo(0L));
+            });
+        }
+        else
+        {
+            perform(checkOptions, AxeCheckOptions.class, options ->
+            {
+                List<AxeReportEntry> report = accessibilityTestExecutor.execute(accessibilityEngine, options,
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:113]:
```suggestion
                List<AxeReportEntry> reportEntries = accessibilityTestExecutor.execute(accessibilityEngine, options,
```

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 10]
```diff
@@ -83,21 +90,54 @@ public AccessibilitySteps(AccessibilityTestEngine accessibilityTestEngine,
...
+                        AxeReportEntry.class);
+
+                attachmentPublisher.publishAttachment(
+                    "/org/vividus/accessibility/axe-accessibility-report.ftl",
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:117]: move to AccessibilityEngine?

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 11]
```diff
@@ -110,4 +150,9 @@ private void publishAttachment(URI pageUrl, List<AccessibilityViolation> accessi
         attachmentPublisher.publishAttachment("/org/vividus/accessibility/violations-table.ftl",
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:150]:
```suggestion
        attachmentPublisher.publishAttachment("/org/vividus/accessibility/htmlcs-accessibility-report.ftl",
```
?

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 12]
```diff
@@ -83,21 +90,54 @@ public AccessibilitySteps(AccessibilityTestEngine accessibilityTestEngine,
...
+        {
+            perform(checkOptions, AxeCheckOptions.class, options ->
+            {
+                List<AxeReportEntry> report = accessibilityTestExecutor.execute(accessibilityEngine, options,
+                        AxeReportEntry.class);
+
+                attachmentPublisher.publishAttachment(
+                    "/org/vividus/accessibility/axe-accessibility-report.ftl",
+                    Map.of("entries", report, "url", currentUrl, "run", options.getRunOnly()),
+                    "Accessibility report for page: " + currentUrl
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:119]:
```suggestion
                    "Accessibility testing report for page: " + currentUrl
```

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java [hunk 13]
```diff
@@ -110,4 +150,9 @@ private void publishAttachment(URI pageUrl, List<AccessibilityViolation> accessi
         attachmentPublisher.publishAttachment("/org/vividus/accessibility/violations-table.ftl",
                 result, "Accessibility violations at page " + pageUrl);
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/AccessibilitySteps.java:151]:
```suggestion
                result, "Accessibility testing report for page: " + pageUrl);
```

File: vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/model/axe/Enabled.java [hunk 14]
```diff
@@ -14,14 +14,19 @@
  * limitations under the License.
  */
 
-package org.vividus.accessibility.engine;
+package org.vividus.accessibility.model.axe;
 
-import java.util.List;
+public class Enabled
```
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/java/org/vividus/accessibility/model/axe/Enabled.java:19]: shouldn't it be noun?

File: vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl
Comment (valfirst) [inline vividus-plugin-accessibility/src/main/resources/org/vividus/accessibility/axe-accessibility-report.ftl:1]: need to add a link to the source of this report

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-allure-adaptor/src/main/resources/allure-customization/webjars/js-beautify/1.13.4/styles/stackoverflow-light.min.css (hunk 6 above)
Comment (uarlouski) [reply vividus-allure-adaptor/src/main/resources/allure-customization/webjars/js-beautify/1.13.4/styles/stackoverflow-light.min.css:3]: image
is contains styles for highlighting

**Comment by valfirst** [comment]
it makes sense to fix the code smell

File: vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    public AxeCheckOptions convertValue(Parameters row, Type type)
+    {
+        String standard = getStandard(row, String.class);
+        List<String> violationsToCheck = getViolationsToCheck(row);
+
+        Validate.isTrue(standard != null || violationsToCheck != null,
+                "Either 'standard' or 'violationsToCheck' must be set");
+
+        Validate.isTrue(standard == null || violationsToCheck == null,
+                "Either 'standard' or 'violationsToCheck' is allowed");
```
Comment (ikalinin1) [inline vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java:50]:
```suggestion
        Validate.isTrue(standard == null ^ violationsToCheck == null,
                "Either 'standard' or 'violationsToCheck' must be set");
```

Comment (uarlouski) [reply vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java:50]: its shorter but in this case we trade off error message, not sure if its worth it

File: .github/workflows/gradle.yml [hunk 16]
```diff
@@ -200,7 +200,8 @@ jobs:
                                                   -Pvividus.notifications.mail.password=${GMAIL_PASSWORD} \
                                                   -Pvividus.notifications.mail.from=vividus.team@gmail.com \
                                                   -Pvividus.notifications.mail.recipient="vividus.team@gmail.com, ivan@vividus.dev" \
-                                                  -Pvividus.notifications.mail.security-protocol=SSL
+                                                  -Pvividus.notifications.mail.security-protocol=SSL \
+                                                  -Pvividus.accessibility.engine=AXE_CORE
```
Comment (valfirst) [inline .github/workflows/gradle.yml:204]: please move this config to the project properties

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

File: vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java (hunk 15 above)
Comment (ikalinin1) [reply vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java:50]: You really don't care about specific messages, you just need to say to user pick either one

**Review by ikalinin1 — APPROVED** [review]

File: vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java (hunk 15 above)
Comment (uarlouski) [reply vividus-plugin-accessibility/src/main/java/org/vividus/converter/ParametersToAxeCheckOptionsConverter.java:50]: on standup It was agreed to keep it as is :)

**Pull request merged**